include/simics/dmllib.h

```diff
@@ -183,6 +235,56 @@ DML_eq(uint64 a, uint64 b)
         ({_traitref_t __tref = traitref;                                \
           ((struct _ ## type *) __tref.trait)->method(dev, __tref);})   \
 
+// A parameter is represented in the vtable using a pointer that points to the
+// parameter value. A parameter whose value varies across indices is stored as
+// an array of all values, indexed by encoded_index. This is marked by setting
+// bit 0 of the vtable's pointer to 1. A parameter that stays constant across
+// indices is represented as a single value, this is marked by setting bit 0 of
+// the pointer to 0. Parameter values are always stored with an alignment of at
+// least 2.
+#define VTABLE_PARAM(traitref, vtable_type, member)                     \
+        ({_traitref_t __tref = traitref;                                     \
+          uintptr_t __member                                                 \
+              = (uintptr_t)((vtable_type *)__tref.trait)->member;            \
+          (__member & 1)                                                     \
+           ? ((typeof(((vtable_type*)NULL)->member))(__member - 1))[         \
+                   __tref.id.encoded_index]                                  \
+           : *(typeof(((vtable_type*)NULL)->member))__member; })
```

Comment:
Perhaps:
```
((typeof(((vtable_type*)NULL)->member))(__member & ~1))[__tref.id.encoded_index * (__member & 1)]
```
?

Reply:
I think what we want is to get a CMOV in assembler. I'll check how to make gcc emit that, another alternative is `((typeof(((vtable_type*)NULL)->member))(__member & ~1))[(__member & 1) ? __tref.id.encoded_index : 0]`

Reply:
I tried the three different options by compiling the following with `gcc -O2`:
```
header %{
extern void chicken(uint64 i);
typedef _traitref_t moo;
%}
template moo { param pig : uint64; }
method horse(moo cow) { chicken(cow.pig); }
export horse as "horse";
```
Current dmllib gives:
```
 8c3:   a8 01                   test   $0x1,%al
 8c5:   75 09                   jne    8d0 <_DML_M_horse.isra.6+0x10>
 8c7:   48 8b 38                mov    (%rax),%rdi
 8ca:   e9 00 00 00 00          jmpq   8cf <_DML_M_horse.isra.6+0xf>
 8cf:   90                      nop
 8d0:   89 f6                   mov    %esi,%esi
 8d2:   48 8b 7c f0 ff          mov    -0x1(%rax,%rsi,8),%rdi
 8d7:   e9 00 00 00 00          jmpq   8dc <_DML_M_horse.isra.6+0x1c>
```
Your suggestion gives:
```
 8a2:   48 83 e0 fe             and    $0xfffffffffffffffe,%rax
 8a6:   83 e1 01                and    $0x1,%ecx
 8a9:   48 0f af d1             imul   %rcx,%rdx
 8ad:   48 8b 3c 02             mov    (%rdx,%rax,1),%rdi
 8b1:   e8 00 00 00 00          callq  8b6 <horse+0x26>
```
My suggestion gives:
```
 8a4:   a8 01                   test   $0x1,%al
 8a6:   48 0f 44 d1             cmove  %rcx,%rdx
 8aa:   48 83 e0 fe             and    $0xfffffffffffffffe,%rax
 8ae:   48 8b 3c 10             mov    (%rax,%rdx,1),%rdi
 8b2:   e8 00 00 00 00          callq  8b7 <horse+0x27>
```